util/sst_file_manager_impl.cc

```diff
@@ -166,6 +231,149 @@ uint64_t SstFileManagerImpl::GetTotalTrashSize() {
   return delete_scheduler_.GetTotalTrashSize();
 }
 
+void SstFileManagerImpl::ReserveDiskBuffer(uint64_t size,
+                                           const std::string& path) {
+  MutexLock l(&mu_);
+
+  reserved_disk_buffer_ += size;
+  if (path_.empty()) {
+    path_ = path;
+  }
+}
+
+void SstFileManagerImpl::ClearError() {
+  while (true) {
+    MutexLock l(&mu_);
+
+    if (closing_) {
+      return;
+    }
+
+    uint64_t free_space;
+    Status s = env_->GetFreeSpace(path_, &free_space);
+    if (s.ok()) {
+      // In case of multi-DB instances, some of them may have experienced a
+      // soft error and some a hard error. In the SstFileManagerImpl, a hard
+      // error will basically override previously reported soft errors. Once
+      // we clear the hard error, we don't keep track of previous errors for
+      // now
+      if (bg_err_.severity() == Status::Severity::kHardError) {
+        if (free_space < reserved_disk_buffer_) {
+          ROCKS_LOG_ERROR(logger_, "free space [%d bytes] is less than "
+              "required disk buffer [%d bytes]\n", free_space,
+              reserved_disk_buffer_);
+          ROCKS_LOG_ERROR(logger_, "Cannot clear hard error\n");
+          s = Status::NoSpace();
+        }
+      } else if (bg_err_.severity() == Status::Severity::kSoftError) {
+        if (free_space < free_space_trigger_) {
+          ROCKS_LOG_WARN(logger_, "free space [%d bytes] is less than "
+              "free space for compaction trigger [%d bytes]\n", free_space,
+              free_space_trigger_);
+          ROCKS_LOG_WARN(logger_, "Cannot clear soft error\n");
+          s = Status::NoSpace();
+        }
+      }
+    }
+
+    // Someone could have called CancelErrorRecovery() and the list could have
+    // become empty, so check again here
+    if (s.ok() && !error_handler_list_.empty()) {
+      cur_instance_ = error_handler_list_.front();
+      // Resume() might try to flush memtable and can fail again with
+      // NoSpace. If that happens, we abort the recovery here and start
+      // over.
+      mu_.Unlock();
+      s = cur_instance_->RecoverFromBGError();
```

Comment:
Is it possible that, before the mutex is acquired again, the same DB recovered, but quickly failed again. Since this DB is still en `error_handler_list_`, it is not inserted again, but we poped it up in line 294, so the error handle of this failure DB is out of track?

Reply:
That's a good catch. I think the safest way for now is to check the DB's error status again after acquiring the mutex, and leave it in the list if there is an error.